semantic/src/main/kotlin/tools/samt/semantic/checks/ValidTypeExpressionCheck.kt

```diff
@@ -0,0 +1,155 @@
+package tools.samt.semantic.checks
+
+import tools.samt.common.DiagnosticConsole
+import tools.samt.parser.*
+import tools.samt.semantic.SemanticCheck
+
+internal class ValidTypeExpressionCheck(private val diagnostics: DiagnosticConsole) : SemanticCheck {
+    override fun check(fileNode: FileNode) {
+        fileNode.statements.forEach {
+            when (it) {
+                is RecordDeclarationNode -> it.fields.forEach { field -> checkTypeExpression(field.type) }
+                is TypeAliasNode -> checkTypeExpression(it.type)
+                is TypeImportNode,
+                is WildcardImportNode,
+                is EnumDeclarationNode,
+                is ConsumerDeclarationNode,
+                is PackageDeclarationNode,
+                is ProviderDeclarationNode,
+                is ServiceDeclarationNode,
+                -> Unit
+            }
+        }
+    }
+
+    private fun checkTypeExpression(expression: ExpressionNode, optionalAllowed: Boolean = true) {
+        when (expression) {
+            is IdentifierNode,
+            is BundleIdentifierNode,
+            -> Unit
+
+            is OptionalDeclarationNode -> {
+                if (optionalAllowed) {
+                    checkTypeExpression(expression.base, false)
+                } else {
+                    diagnostics.reportError("Cannot nest optional types", expression.location)
+                }
+            }
+
+            is CallExpressionNode -> {
+                checkConstraintBaseTypeExpression(expression.base)
+                for (argument in expression.arguments) {
+                    checkConstraintArgumentExpression(argument)
+                }
+            }
+
+            is GenericSpecializationNode -> diagnostics.reportError(
+                "Generic type are currently not supported",
```

Comment:
Move this into the `NotImplementedFeaturesCheck` pass

Reply:
The problem is then I'll have to also implement the expressiont traversial over there. Alternatively a visitor implementation would reduce this duplication, but it seems quite overkill for our simple checks, leaving the generics here for the time being was the simples approach. I don't believe it's worth the effort to separate at the moment, what's your take?